### PR TITLE
ENH: optimize: add the Hungarian algorithm

### LINKED ISSUE #4482: wishlist: O(N^3) solution of weighted assignment problem
SciPy has a maximum bipartite matching implementation in csgraph, but it is unweighted.

See also http://en.wikipedia.org/wiki/Hungarian_algorithm.
And a thread on the mailing list: http://mail.scipy.org/pipermail/scipy-user/2013-March/034375.html

### PR DESCRIPTION
**Pull request by larsmans**
The Kuhn-Munkres (Hungarian) algorithm for solving the linear-sum assignment problem.
- <del>Needs more documentation. In particular, the meaning of the rectangular input case should be documented. I'm not even sure what it does.</del>
- <del>Needs an appropriate function name. Should this be called this after the problem it solves (`linear_assignment`) or after the algorithm (`hungarian` or `kuhn_munkres`)? There are several algorithms to solve this problem and the implementation could be changed to a different one later.</del>

Fixes #4482, fixes scikit-learn/scikit-learn#5103.

### CODE REVIEW

**Comment by larsmans** [comment]
Btw., I tried to keep the commit history from scikit-learn, but the files had been `git mv`'d twice which makes this near impossible.

**Comment by larsmans** [comment]
@bmc might be interested to know that his code ends up here.

**Comment by bmc** [comment]
Oh, hilarious... Of course, I just adapted the code from a Perl implementation. And this version has been optimized, so it's better than my original. Thanks for the mention, though. I'll modify my standalone implementation to point to this one.

**Comment by charris** [comment]
Nice to see this.  I'd vote for naming after the algorithm as there are several available. A good cross reference to linear assignment in the documentation would help users find the relevant function.

**Comment by rgommers** [comment]
> I'd vote for naming after the algorithm as there are several available

Isn't that a reason to name it after the problem it solves? Otherwise you'd have to rename it later if a strictly better algorithm comes along. Or you have several in parallel.

I think I'd rather have ``linear_assignment`` and when a second one comes along ``linear_assignment(method='hungarian')``.

**Comment by rgommers** [comment]
Test error is real: 

    TypeError: asarray() got an unexpected keyword argument 'copy'

**Comment by charris** [comment]
Good point.

**Comment by larsmans** [comment]
So naming after the problem it is? Then I'll rename the function `linear_sum_assignment` because there are actually several linear assignment problems.

Should I include a `method` parameter for forward compatibility?

**Comment by rgommers** [comment]
> Should I include a method parameter for forward compatibility?

You don't need that, a new keyword can always be added later with default ``'hungarian'``. There's then no difference in fw compat.

**Comment by larsmans** [comment]
Ok, I just did a pretty big overhaul of the code: it now returns the solution as `row_ind, col_ind` (just like `np.where`), which is easier to work with. Previously, you had to do `cost_matrix[np.ix_(*result)]` to reconstruct the assignment costs. I also extended the docs to cover the rectangular case, renamed the function as discussed and changed the tests to check the whole path of costs rather than just the total.

**Comment by ewmoore** [comment]
Looks like there are a number of little things here.  

1.  It chokes if given a list or `np.matrix`
2.  Would be nice to include an example.
3. `.. versionadded:: 0.17.0` under notes.

**Comment by larsmans** [comment]
@ewmoore Fixed.

**Comment by ewmoore** [comment]
I'm sorry I missed this before, but could you add this to the release notes?  I think its ready to go then.

**Comment by larsmans** [comment]
@ewmoore done.

**Comment by ewmoore** [comment]
Great, thanks @larsmans!

**Comment by jnothman** [comment]
How hard would it be to adapt this to the case where there are only sparsely many finite scores in the matrix (or nonzeros where all nonzeros are negative)? To what extent might we expect speed-up in this case (assuming that the finite scores still produce a single connected component rather than separable problems).

File: scipy/optimize/tests/test_hungarian.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        # n == 2, m == 0 matrix
+        ([[], []],
+         []),
+    ]:
+        cost_matrix = np.array(cost_matrix)
+        row_ind, col_ind = linear_sum_assignment(cost_matrix)
+        assert_array_equal(row_ind, np.sort(row_ind))
+        assert_array_equal(expected_cost, cost_matrix[row_ind, col_ind])
+
+        row_ind, col_ind = linear_sum_assignment(cost_matrix.T)
```
Comment (jnothman) [inline scipy/optimize/tests/test_hungarian.py]: More importantly, surely this should be `col_ind, row_ind`??? The scikit-learn version includes logic to transpose the input and output if it is a tall matrix... why did that disappear?

Comment (larsmans) [inline scipy/optimize/tests/test_hungarian.py]: Hm, yes, you're right. I forgot to transpose the `cost_matrix` below when writing this test.

**Pull request merged**